### PR TITLE
DNN backends registry

### PR DESCRIPTION
**Pull request by mshabunin**
### This pullrequest changes

adds basic registry for DNN backends and targets

```
docker_image:Custom=ubuntu-openvino:16.04
buildworker:Custom=linux-2
test_opencl:Custom=OFF
```

### CODE REVIEW

File: modules/dnn/test/test_caffe_importer.cpp [hunk 1]
```diff
@@ -324,7 +324,8 @@ TEST_P(Reproducibility_SqueezeNet_v1_1, Accuracy)
     Mat ref = blobFromNPY(_tf("squeezenet_v1.1_prob.npy"));
     normAssert(ref, out);
 }
-INSTANTIATE_TEST_CASE_P(/**/, Reproducibility_SqueezeNet_v1_1, availableDnnTargets());
+INSTANTIATE_TEST_CASE_P(/**/, Reproducibility_SqueezeNet_v1_1,
+    testing::ValuesIn(getAvailableTargets(DNN_BACKEND_OPENCV)));
```
Comment (dkurt) [inline modules/dnn/test/test_caffe_importer.cpp]: It enables FP16 tests so we can use common thresholds:

```cpp
double l1 = (target == DNN_TARGET_OPENCL_FP16 || target == DNN_TARGET_MYRIAD) ? 4e-3 : 1e-5;
double lInf = (target == DNN_TARGET_OPENCL_FP16 || target == DNN_TARGET_MYRIAD) ? 2e-2 : 1e-4;
```
It should fix the following tests:
```
Reproducibility_SqueezeNet_v1_1.Accuracy/1, where GetParam() = OCL_FP16
Reproducibility_GoogLeNet.Batching/1, where GetParam() = OCL_FP16
```
However the following ones doesn't pass:
```
Reproducibility_GoogLeNet.IntermediateBlobs/1, where GetParam() = OCL_FP16
Reproducibility_GoogLeNet.SeveralCalls/1, where GetParam() = OCL_FP16
```
I think it's OK if we skip them for now for OCL_FP16 target.

File: modules/dnn/test/test_backends.cpp [hunk 2]
```diff
@@ -292,6 +292,6 @@ TEST_P(DNNTestNetwork, FastNeuralStyle_eccv16)
     processNet("dnn/fast_neural_style_eccv16_starry_night.t7", "", inp, "", "", l1, lInf);
 }
 
-INSTANTIATE_TEST_CASE_P(/*nothing*/, DNNTestNetwork, dnnBackendsAndTargets(true, true, false));
+INSTANTIATE_TEST_CASE_P(/*nothing*/, DNNTestNetwork, dnnBackendsAndTargets());
```
Comment (dkurt) [inline modules/dnn/test/test_backends.cpp]: I think we need to keep `false` for `withCpuOCV`

File: modules/dnn/test/test_common.hpp [hunk 3]
```diff
@@ -237,49 +213,31 @@ namespace opencv_test {
 using namespace cv::dnn;
 
 static inline
-testing::internal::ParamGenerator<tuple<Backend, Target> > dnnBackendsAndTargets(
+testing::internal::ParamGenerator< tuple<Backend, Target> > dnnBackendsAndTargets(
         bool withInferenceEngine = true,
         bool withHalide = false,
         bool withCpuOCV = true
```
Comment (dkurt) [inline modules/dnn/test/test_common.hpp]: `withCpuOCV` is ignored here. Some tests use default CPU implementation as reference so we don't compare it with itself.

**Comment by alalek** [comment]
I will prepare fixup commit for that.

**Pull request merged**